File: Code/matlab/KalmanFilter.m
function [MU, SIGMA] = KalmanFilter(U, Z, A, B, C, Q, R, mu0, Sigma0)

%%
% 
% PURPOSE
% --------------
% Running a full Kalman filtering round for the traditional state space model:
% 
% z(t) = C*x(t) + delta(t), delta(t) ~ N(0,Q_(t))             Measurement equation
% x(t) = A*x(t-1) + B*u(t) + eps(t), eps(t) ~ N(0,R_(t))      State equation
% where x(t) is the n-dim state, u(t) is the n-dim control, z(t) is the k-dim observed data.
% 
%
% CALL
% --------------
% [mu, Sigma] = KalmanFilter(mu, Sigma, u, z, A, B, C, Q, R)
%        
% INPUTS
% --------------
% U                 T-by-m          U(t,:) is u(t)'             
% Z                 T-by-k          Z(t,:) is z(t)'
% A, B, C, Q, R                     Model parameters, see the state-space model above.
% mu0               n-by-1          Initial mean of state
% Sigma0            n-by-n          Initial covariance of state
%
% OUTPUTS
% ---------------
% MU                T-by-n          MU(t,:) is the posterior mean of the state at time t                 
% SIGMA             n-by-n-by-T     SIGMA(:,:,t) is the posterior covariance matrix of the state at time t
%
% AUTHOR
% ---------------
% Mattias Villani, Link?ping University. e-mail: mattias.villani@gmail.com
%
% VERSION DATING
% ---------------
% FIRST     2015-07-29
% CURRENT   2015-07-29
%
% REFERENCES
% ---------------
% Thrun, Burgard and Fox (2006). Probabilistic Robotics, Algorithm Kalman_filter in Table 3.1.

%% Prelims
T = size(Z,1); 
n = length(mu0);

MU = zeros(T,n);
SIGMA = zeros(n,n,T);

% The Kalman iterations
mu = mu0;
Sigma = Sigma0;
for t = 1:T
    [mu, Sigma] = KalmanFilterUpdate(mu, Sigma, U(t,:)', Z(t,:)', A, B, C, Q, R);
    MU(t,:) = mu;
    SIGMA(:,:,t) = Sigma;
end

File: Code/julia/KalmanFilter.jl
""" 
    kalmanfilter_update(μ, Σ, u, y, A, B, C, Q, R) 

A single Kalman filter update at time t of the state space model: 

yₜ = Cxₜ + δₜ,           δₜ ~ N(0,Qₜ)         Measurement equation
xₜ = Axₜ₋₁+ Buₜ + εₜ,    εₜ ~ N(0,Rₜ)         State equation

where
xₜ is the n-dim state
uₜ is the n-dim control
yₜ is the k-dim observed data. 

Reference: Thrun, Burgard and Fox (2006). Probabilistic Robotics, Algorithm Kalman_filter in Table 3.

"""
function kalmanfilter_update(μ, Σ, u, y, A, B, C, Q, R)

    # Prediction step - moving state forward without new measurement
    μ̄ = A*μ + B*u;
    Σ̄ = A*Σ*A' + R;

    # Measurement update - updating the N(μ̄, Σ̄) prior with the new data point
    K = Σ̄*C' / (C*Σ̄*C' + Q); # Kalman Gain
    μ = μ̄ + K*(y - C*μ̄);
    Σ = ( I(length(μ)) - K*C )*Σ̄;

    return μ, Σ
end

""" 
    kalmanfilter(U, Y, A, B, C, Q, R, μ₀, Σ₀) 

A full run of the Kalman filter update at time t of the state space model: 

yₜ = Cxₜ + δₜ,           δₜ ~ N(0,Qₜ)         Measurement equation
xₜ = Axₜ₋₁+ Buₜ + εₜ,    εₜ ~ N(0,Rₜ)         State equation

where
xₜ is the n-dim state
uₜ is the n-dim control
yₜ is the k-dim observed data. 

The observed data observations are the rows of the T×k matrix Y
The control signals are the rows of the T×m matrix U
μ₀ and Σ₀ are the mean and covariance of the initial state vector x₀

Reference: Thrun, Burgard and Fox (2006). Probabilistic Robotics, Algorithm Kalman_filter in Table 3.

"""

function kalmanfilter(U, Y, A, B, C, Q, R, μ₀, Σ₀)
    
    # Prelims
    T = size(Y,1)
    n = length(μ₀)
    
    μ_all = zeros(T,n)
    Σ_all = zeros(n,n,T)
    
    # The Kalman iterations
    μ = μ₀
    Σ = Σ₀
    for t = 1:T
        μ, Σ = kalmanfilter_update(μ, Σ, U[t,:]', Y[t,:]', A, B, C, Q, R)
        μ_all[t,:] = μ
        Σ_all[:,:,t] = Σ
    end

    return μ_all, Σ_all
end
    